Project and session goals

BlayW opened the pull request:
**Summary:**

I've added project settings for project word count goal, project completion target date, and a session word count goal that can optionally be automatically calculated based off the project completion target date. These are also visualised in the status bar as progress bars that update alongside the existing word count label in the status bar.

**Related Issue(s):**

Closes #2223

**Reviewer's Checklist:**

* [x] The header of all files contain a reference to the repository license
* [ ] The overall test coverage is increased or remains the same as before
* [ ] All tests are passing
* [ ] All flake8 checks are passing and the style guide is followed
* [x] Documentation (as docstrings) is complete and understandable
* [x] Only files that have been actively changed are committed

BlayW committed:
added session and project goal with progress bars for each in the status bar

vkbo commented:
Great! This is a good start.

Here are some ToDos:

* I prefer to design a custom progress bar widget instead of relying on style sheets. Style sheets are annoying to deal with in terms of theme updates, so I only use them when there are no other options. I can create a new progress bar widget for this purpose. There are already two custom ones in the code, so it's pretty straightforward to duplicate and tweak.
* I think this is a complex enough feature to warrant its own core class. If not for anything else other than handling the settings internally. Which means there would only be a single value added to NWProjectData.
* The progress bars should be hidden if the goals are set to 0. I'm unsure whether 0 should be the default though.

I can handle the custom widget and the core class if you wish, or we can split them. The core class should be modelled after the NWStatus class. You don't need to use the NW prefix. I don't use those any more, and plan to remove them. ProjectGoal is a sensible name.

Some implementation comments:

* In this project it is preferred to use Python native features before the Qt ones. Since the Qt library is implemented in C++ with a Python interface layer and two competing garbage collectors, it is simpler. So, just use the Python date module instead of QDate.
* The PR will be run through flake8 linting with some rules and restrictions, so it's a good idea to also run it locally to avoid linting errors. I saw a few. Not sure what editor you use, but VSCode has flake8 support built into its Python extension.
* Anything added to the project need to be 100% covered by tests. But we can handle that when everything else is done.

BlayW commented:
I initially considered adapting one of the existing custom progress bars, so I'll create the widget as a start, at least.

BlayW commented:
Is it okay with you if I change a parent class for `NProgressCircle` that contains everything except `paintEvent`, it will reduce duplicate code I think.

vkbo commented:
> Is it okay with you if I change a parent class for `NProgressCircle` that contains everything except `paintEvent`, it will reduce duplicate code I think.

Just duplicate for now, and it it turns out they do not diverge, we can create the parent class later. I'm planning a different colouring feature for this progress bar, so I don't think they will stay that similar.

BlayW committed:
converted QDate to datetime.date
Added NProgressGoal class to handle goal progress bars.

BlayW committed:
Merge branch 'main' into project-goals

BlayW commented:
My most recent push adds the new `NProgressGoals` progress bar class while retaining most of the control in the `status bar.py` file. I expect a lot of that functionality will be moved to the new core class, so I tried not to alter it at all beyond retaining the current functionality. I also removed the `QDate` usage in favour of `date time.date` objects and addressed the flake8 issues

vkbo commented on .vscode/settings.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "python.testing.pytestArgs": [
+        "tests"
+    ],
+    "python.testing.unittestEnabled": false,
+    "python.testing.pytestEnabled": true
+}
```
Please do not commit local config files to the repo.

BlayW replied:
whoops, sorry!

vkbo commented on novelwriter/dialogs/projectsettings.py:
```diff
@@ -353,7 +354,8 @@ def __init__(self, parent: QWidget) -> None:
         self.addRow(
             self.tr("Auto populate session goal?"), self.sessGoalAuto,
-            self.tr("Calculate Session goal based on target date assuming daily sessions. Overrides configured session goal below.")
+            self.tr("Calculate Session goal based on target date assuming daily sessions. \
+                    Overrides configured session goal below.")
```
Please don't use backslash for line continuation. This is the correct way:
```suggestion
            self.tr(
                "Calculate Session goal based on target date assuming daily sessions. "
                "Overrides configured session goal below."
            )
```

vkbo commented:
> My most recent push adds the new `NProgressGoals` progress bar class while retaining most of the control in the `status bar.py` file. I expect a lot of that functionality will be moved to the new core class, so I tried not to alter it at all beyond retaining the current functionality. I also removed the `QDate` usage in favour of `date time.date` objects and addressed the flake8 issues

Thanks! I'll take a look at the branch a bit later and add the data class for the config.

I'm still wondering if the progress bars should be in the side panel rather than the status bar, but I will create an example and we can compare from screenshots. I need to see it side by side to get a good impression.

BlayW committed:
Update novelwriter/dialogs/projectsettings.py

Co-authored-by: Veronica Berglyd Olsen <1619840+vkbo@users.noreply.github.com>

BlayW committed:
removed local config files

vkbo commented:
Sorry for the slow progress from my side. I don't have time to work on new code right now as I've been very busy with work and other things. I'll get back to it soon.

vkbo commented:
As you may have seen I've postponed remaining features till 2.8 while I work to get 2.7 done. I'll pick this up again when I start working on 2.8.
